qa/tasks/cephfs/test_exports.py

```diff
@@ -580,3 +580,50 @@ def test_ephemeral_pin_shrink_mds(self):
         log.info("{0} migrations have occured due to the cluster resizing".format(count))
         # rebalancing from 3 -> 2 may cause half of rank 0/1 to move and all of rank 2
         self.assertLessEqual((count/len(subtrees_old)), (1.0/3.0/2.0 + 1.0/3.0/2.0 + 1.0/3.0)*1.25) # aka .66 with 25% overbudget
+
+class TestKillExports(CephFSTestCase):
+    MDSS_REQUIRED = 3
+    CLIENTS_REQUIRED = 1
+
+    def setUp(self):
+        CephFSTestCase.setUp(self)
+
+        self.fs.set_max_mds(self.MDSS_REQUIRED)
+        self.status = self.fs.wait_for_daemons()
+
+        self.mount_a.run_shell_payload('mkdir -p test/export')
+
+    def tearDown(self):
+        super().tearDown()
+
+    def _kill_export(self, rank, kill):
+        self.fs.rank_asok(['config', 'set', 'mds_kill_export_at', str(kill)], rank=rank, status=self.status)
+
+    def _export_dir(self, path, source, target):
+        self.fs.rank_asok(['export', 'dir', path, str(target)], rank=source, status=self.status)
+
+    def _wait_laggy(self, rank):
+        self.wait_until_true(lambda: "laggy_since" in self.fs.get_rank(rank=rank), timeout=self.fs.beacon_timeout)
+
+    def _clear_coredump(self, rank):
+        crash_rank = self.fs.get_rank(rank=rank, status=self.status)
+        self.delete_mds_coredump(crash_rank['name'])
+
+    def test_clear_session(self):
+        source = 0
+        target = 1
+        kill = 9
```

Comment:
Importantly to confirm this fix: also test `kill = 11`.

Reply:
> comment this magic number

I used more human readable variable names.

Reply:
I think patrick meant to test both 9 and 11? I see its now changed to 11

Reply:
> Importantly to confirm this fix: also test `kill = 11`.

I add a new test case for it, but it cannot trigger this bug.

Reply:
ah, the code is repetitive, maybe you can test both 9 and 11 in the same test case itself, something like this?
```
        kill_export_at = [9, 11] # can we add more to test for better code coverage?

        exporter_rank = 0
        importer_rank = 1

        for val in kill_export_at:
	        self._run_kill_export(val, exporter_rank, importer_rank)

	        client_id = self.mount_a.get_global_id()
	        self.mount_a.umount_wait()

	        # timeout if buggy
	        self.wait_until_evicted(client_id, importer_rank)
```

Reply:
> I think patrick meant to test both 9 and 11? I see its now changed to 11

I'm not sure if we still need to test `mds_kill_export_at = 11` because it cannot trigger this bug. This test case will crash the mds because of my wrong modification in the previous modification, but now it won't.

Reply:
I think it is always good to have better code coverage, i.e. testing it with all possible questionable values (even if chances of them leading to a bug maybe slim).

Reply:
what do you think @batrick ?

Reply:
> Sorry. I think I meant `10` not `11`. That would trigger the importer to reverse the import and test your change.

https://github.com/ceph/ceph/blob/d146d2b56c6fd33eb52a85349b04080b31ee06c4/src/mds/MDCache.cc#L3270-L3276

`import_reverse` will be called if `mds_kill_export_at` be set to `9`, so it can trigger the bug in `import_reverse`.
`import_finish` will be called if `mds_kill_export_at` be set to `10`, because `EExport` has already been logged.
So setting `mds_kill_export_at` to 10 will not trigger this bug too, and do we need to add a test for it?

Reply:
https://github.com/ceph/ceph/blob/227ea203b006c774b05ab2718e115a956bd7bb49/src/mds/Migrator.cc#L1928

The event will not be persisted yet by the time this assert is hit (the assert really should be _before_ the journal entry is submitted, otherwise it's racy). Since `EExport` will not be logged by the exporter, the importer should be forced to reverse the import during recovery.

Reply:
> https://github.com/ceph/ceph/blob/227ea203b006c774b05ab2718e115a956bd7bb49/src/mds/Migrator.cc#L1928
> 
> The event will not be persisted yet by the time this assert is hit (the assert really should be _before_ the journal entry is submitted, otherwise it's racy). Since `EExport` will not be logged by the exporter, the importer should be forced to reverse the import during recovery.

I thought so too until I tested it in my test environment. I found the replay log of `EExport` in the exporter's log after the test was done, which means that `EExport` had been logged. I had tested it multiple times and got the same result every time. I added sleep code in `MDLog::_submit_thread` and tested it again, the replay log of `EExport` would no longer be found in the exporter's log.

Reply:
> (You should fix that in this PR!)

OK, I've added a new commit to fix it.